substrate/frame/migrations/src/lib.rs

```diff
@@ -785,4 +785,100 @@ impl<T: Config> MultiStepMigrator for Pallet<T> {
 	fn step() -> Weight {
 		Self::progress_mbms(System::<T>::block_number())
 	}
+
+	#[cfg(feature = "try-runtime")]
+	fn try_mbms() -> Result<(), sp_runtime::TryRuntimeError> {
+		use sp_std::collections::btree_map::BTreeMap;
+
+		// Check if migrations were initiated.
+		if !Self::ongoing() {
+			return Ok(())
+		}
+
+		// Store pre_upgrade state for all migrations.
+		let mut pre_upgrade_state = BTreeMap::<u32, Vec<u8>>::new();
+		let mut errors = Vec::new();
+		for i in 0..T::Migrations::len() {
+			match T::Migrations::nth_pre_upgrade(i) {
+				Ok(state) => {
+					pre_upgrade_state.insert(i, state);
+				},
+				Err(e) => {
+					errors.push((i, "pre-upgrade", e));
+				},
+			}
+		}
+
+		// Run migrations to completion.
+		match Self::try_step_until_completion() {
+			Ok(blocks) => {
+				log::info!("ℹ️ MBMs executed in {} blocks", blocks);
+			},
+			Err((i, e)) => errors.push((i, "step", e)),
+		}
+
+		// Run post_upgrade for all migrations.
+		for i in 0..T::Migrations::len() {
+			if let Err(e) = T::Migrations::nth_post_upgrade(
+				i,
+				pre_upgrade_state.get(&i).unwrap_or(&vec![]).clone(),
+			) {
+				errors.push((i, "post-upgrade", e));
+			}
+		}
+
+		if !errors.is_empty() {
+			log::error!("❌ MBM errors occured:");
+			for (i, p, e) in errors {
+				log::error!(
+					"  - Migration at index {} errored during phase '{}' error: {:?}",
+					i,
+					p,
+					e
+				);
+			}
+			return Err("MBM failure/s. See error logs for info.".into())
+		}
+
+		Ok(())
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	#[cfg(feature = "try-runtime")]
+	fn try_step_until_completion() -> Result<BlockNumberFor<T>, (u32, sp_runtime::TryRuntimeError)>
+	{
+		use frame_support::traits::Hooks;
+
+		let start_block = System::<T>::block_number();
+		loop {
+			let migration_index = match Cursor::<T>::get() {
+				Some(MigrationCursor::Active(active)) => active.index,
+				_ => unreachable!("migration must be active"),
+			};
+
+			// Simulate a step.
+			log::debug!("Block {}", System::<T>::block_number());
+			System::<T>::set_block_number(System::<T>::block_number() + 1u32.into());
+			System::<T>::on_initialize(System::<T>::block_number());
+			crate::Pallet::<T>::on_initialize(System::<T>::block_number());
+			Self::step();
+			<crate::Pallet<T>>::on_finalize(System::<T>::block_number());
+			<frame_system::Pallet<T>>::on_finalize(System::<T>::block_number());
+
+			// Check events for `UpgradeCompleted` or `UpgradeFailed`.
+			for e in System::<T>::events().iter() {
```

Comment:
```suggestion
			for e in frame_system::Events::<T>::take().into_iter() {
```
Should make it a bit faster to not hold old events.